### PR TITLE
small naming changes and performance improvement through memoization …

### PR DESCRIPTION
**Pull request by mkleinegger**
performance improvements and naming changes; not fully finished, docs are currently added and small code chunk is in there to be tested

### CODE REVIEW

File: monte_carlo.py [hunk 1]
```diff
@@ -79,26 +80,27 @@ def __init__(self, grid, gamma=0.9, num_episodes=10000, train_epsilon=0.9, test_
...
-            if np.all(np.abs(combined) <= MonteCarlo.max_velocity) and np.any(combined):
-                legal_actions.append(c)
-
-        legal_actions = [list(MonteCarlo.actions.keys())[list(MonteCarlo.actions.values()).index(a)] for a in legal_actions]
-        return legal_actions
+    def get_valid_actions(state):
+        valid_actions = []
+        
+        if state not in MonteCarlo.actions_memoization:
+            for action, step in MonteCarlo.actions.items():
```
Comment (max-seeli) [inline monte_carlo.py]:
```suggestion        
        if state not in MonteCarlo.actions_memoization:
            valid_actions = []
            for action, step in MonteCarlo.actions.items():
```

I think this scope makes more sense

File: monte_carlo.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import numpy as np
+import pandas as pd
```
Comment (max-seeli) [inline monte_carlo.py]:
```suggestion
```

I think this is not used?

**Review by max-seeli — COMMENTED** [review]
Great improvements, especially memoization seems to work great!

File: monte_carlo.py [hunk 3]
```diff
@@ -43,7 +44,7 @@ def save(self, filename):
                 f.write(f"{state[0]},{state[1]},{state[2]},{state[3]},{action}\n")
     
 class MonteCarlo:
-
+    actions_memoization = {}
```
Comment (rebeccavfweiss) [inline monte_carlo.py]: maybe rename as 
valid_actions_memoization
for immediate understanding as I had to see changes in get_valid_actions() to see what it is used for

File: monte_carlo.py (hunk 1 above)
Comment (mkleinegger) [reply monte_carlo.py]: i though about this too; but was questioning about making the `action_memoization` a `defaultdict(list)`; but on second sight, i think this is a more self explanatory implementation

File: monte_carlo.py [hunk 4]
```diff
@@ -7,43 +7,108 @@
 
 
 class Policy:
+    """
+    A wrapper class for the policy dictonary of the `MonteCarlo` agent.
+    """
+
     def __init__(self, train_epsilon, test_epsilon):
+        """
+        Initializes an instance of the MonteCarlo Reinforcement Learning agent.
```
Comment (max-seeli) [inline monte_carlo.py]:
```suggestion
        Initializes an instance of the MonteCarlo Reinforcement Learning policy for an agent.
```

File: monte_carlo.py [hunk 5]
```diff
@@ -7,43 +7,108 @@
...
+
+        Parameters
+        ----------
+        state: tuple(int, int, int, int)
+            The state for which `actions` should be set.
+        actions: [str]
+            The actions which should be associated to the specified `state`.
+        """
+
+        self.policy[state] = actions
```
Comment (max-seeli) [inline monte_carlo.py]:
```suggestion
    def __setitem__(self, state, action):
        """
        Sets the `action` for the specified `state`.

        Parameters
        ----------
        state: tuple(int, int, int, int)
            The state for which `action` should be set.
        action: str
            The action which should be associated to the specified `state`.
        """

        self.policy[state] = action
```

Why would it be several actions?

File: monte_carlo.py [hunk 6]
```diff
@@ -79,26 +162,46 @@ def __init__(self, grid, gamma=0.9, num_episodes=10000, train_epsilon=0.9, test_
...
+                    valid_actions.append(action)
+
+
+            MonteCarlo.valid_actions_memoization[state] = valid_actions
+        return MonteCarlo.valid_actions_memoization[state]
     
     def generate_episode(self, start=None):
-        
+        """
+        Runs an episode of the game. Episode ends with reaching the target cell and gets resetted to an new
```
Comment (max-seeli) [inline monte_carlo.py]:
```suggestion
        Runs an episode of the game. Episode ends with reaching the target cell and gets resetted to a new 
```

File: monte_carlo.py [hunk 7]
```diff
@@ -52,6 +117,24 @@ class MonteCarlo:
...
+        """
+        Initializes an instance of the MonteCarlo Reinforcement Learning agent.
+
+        Parameters
+        ----------
+        grid : Grid
+            The environment grid for the agent to navigate.
+        gamma : float
+            The discount factor for future rewards. Defaults to 0.9.
+        train_epsilon : float
```
Comment (max-seeli) [inline monte_carlo.py]:
```suggestion
            The discount factor for future rewards. Defaults to 0.9.
        num_episodes : int
            The number of episodes the agent is trained. Defaults to 10000
        train_epsilon : float
```

**Review by max-seeli — APPROVED** [review]
Only comment adjustments

**Pull request merged**